### PR TITLE
Issue 1115-Resiliency fix to not crash the app

### PR DESCRIPTION
**Pull request by kreedula**
Added validation checks for authority and clientid when constructing TokenCacheItem from SSOStateSerializer
Flipped the comparision check for authority in isMultipleRTsMatchingGivenAppAndResource method
Unit test cases

### CODE REVIEW

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java [hunk 1]
```diff
@@ -2659,6 +2659,32 @@ public void testDeserializeMissingAttribute() {
...
+
+    @Test
+    public void testDeserializeMissingClientId() {
+        final String missingAttributeString = "{\"tokenCacheItems\":[{\"authority\":\"https://login.windows.net/ComMon/\",\"refresh_token\":\"FRT\"}],\"version\":1}";
+        final FileMockContext mockContext = new FileMockContext(InstrumentationRegistry.getContext());
+        final DefaultTokenCacheStore mockCache = new DefaultTokenCacheStore(mockContext);
+        final AuthenticationContext context = getAuthenticationContext(mockContext, VALID_AUTHORITY, false, mockCache);
+        try {
+            context.deserialize(missingAttributeString);
+        } catch (final Exception exception) {
```
Comment (iambmelt) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java]: Similar comment as above re using `@Test(expected=...)` params

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java [hunk 2]
```diff
@@ -2659,6 +2659,32 @@ public void testDeserializeMissingAttribute() {
...
 
+    @Test
+    public void testDeserializeMissingAuthority() {
+        final String missingAttributeString = "{\"tokenCacheItems\":[{\"refresh_token\":\"FRT\",\"foci\":\"1\"}],\"version\":1}";
+        final FileMockContext mockContext = new FileMockContext(InstrumentationRegistry.getContext());
+        final DefaultTokenCacheStore mockCache = new DefaultTokenCacheStore(mockContext);
+        final AuthenticationContext context = getAuthenticationContext(mockContext, VALID_AUTHORITY, false, mockCache);
+        try {
+            context.deserialize(missingAttributeString);
+        } catch (final Exception exception) {
```
[the remark this answers is not in the document]
Comment (kreedula) [reply adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java]: Thank you for the suggestion Brian. The code works, I have made changes to these tests and also to couple of previous tests in the test suite to ensure we follow this paradigm.

File: adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java (hunk 1 above)
Comment (kreedula) [reply adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java]: Done. Made changes as needed.

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java [hunk 3]
```diff
@@ -1278,6 +1278,10 @@ void deserialize(final String serializedBlob) throws AuthenticationException {
         // first, we'll fall back to passed in authority host and all the aliased hosts if necessary; If app receives the blob has the old
         // version of adal, since all the apps using token share library are using login.windows.net, token lookup will keep working.
         final TokenCacheItem tokenCacheItem = SSOStateSerializer.deserialize(serializedBlob);
+        if (tokenCacheItem == null || StringExtensions.isNullOrBlank(tokenCacheItem.getAuthority()) ||
```
[the remark this answers is not in the document]
Comment (kreedula) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java]: Yes you're right, `tokenCacheItem == null` is not needed as we throw an `AuthenticationException` if it's not passed in. I have added a test case for that scenerio as well. Also also added final to params as required.

**Comment by kreedula** [comment]
Sure. Thanks for the review @iambmelt

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java [hunk 4]
```diff
@@ -1278,6 +1278,10 @@ void deserialize(final String serializedBlob) throws AuthenticationException {
         // first, we'll fall back to passed in authority host and all the aliased hosts if necessary; If app receives the blob has the old
         // version of adal, since all the apps using token share library are using login.windows.net, token lookup will keep working.
         final TokenCacheItem tokenCacheItem = SSOStateSerializer.deserialize(serializedBlob);
+        if (StringExtensions.isNullOrBlank(tokenCacheItem.getAuthority()) ||
```
Comment (heidijinxujia) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java]: Could tokenCacheItem be null?

File: adal/src/main/java/com/microsoft/aad/adal/SSOStateSerializer.java [hunk 5]
```diff
@@ -120,7 +120,7 @@ private String internalSerialize() {
      * @return TokenCacheItem
      * @throws AuthenticationException
      */
-    private TokenCacheItem internalDeserialize(String serializedBlob) throws AuthenticationException {
+    private TokenCacheItem internalDeserialize(final String serializedBlob) throws AuthenticationException {
```
Comment (heidijinxujia) [inline adal/src/main/java/com/microsoft/aad/adal/SSOStateSerializer.java]: 👍

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java (hunk 4 above)
Comment (kreedula) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java]: No, we throw an `AuthenticationException`  if we can't construct the TokenCacheItem. I have added a unit test case here for this specific case : https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/1209/commits/678e24d37df8d44233ddb9e96a691f1a5c8ccbfc#diff-1f51ca351931fc74098edcfa59eecbd3R2630

**Comment by kreedula** [comment]
Thanks @heidijinxujia for the review. Replied to your comment. Let me know what you think

**Pull request merged**